Invoice update

kevinpapst opened the pull request:
FIXES #468 
FIXES #187 

Changes proposed in this pull request:
- styling invoice extension screen
- removed rounding checkbox in favor of an empty option in rounding value select box (UI improvement)
- added failure handling for renderer
- setting customer contact data to empty string if not available
- added 3 select options for entry order: date-asc, date-desc, by name
- validating input parameter
- using own datepicker instead of relying on global timeframe
- new HTML invoice template: freelancer
- group invoice templates by type
- renamed existing invoice templates
- more fields available for rendering invoices
- rewritten and simplified code for future improvements

Changes within the core:
- reusing date_format_0 for datepicker (date_format_0 was not used anymore in Kimai)
- introducing a new date_format_3 setting to allow safely parsing datepicker values in PHP

This allows us to improve the datepicker usage in the future with different locales.

simonschaufi commented on extensions/ki_invoice/init.php:
```diff
@@ -52,6 +52,7 @@
 // Select values for Round Time option
 $roundingOptions = array(
+    '0' => '',
     '1' => '0.1h',
```
I don't like the format of 0.1h. If you calculate 100 as a base (decimal format), then 0.1h is 0.6 min which is not the same as 1 min.
This should be written as 1min, 30min..., then it is clear.

0.25h can also mean 25min but we mean 15min here.

simonschaufi commented:
nice layout! I :heart: it!

simonschaufi commented on extensions/ki_invoice/templates/scripts/main.php:
```diff
@@ -6,45 +6,78 @@
+                    <label for="short">
+                        <?php echo $this->kga['lang']['ext_invoice']['invoiceOptionShort'] ?>
+                    </label>
+                    <input type="checkbox" name="short" checked="checked">
```
id="short" missing to connect the label with the checkbox

simonschaufi commented on extensions/ki_invoice/templates/scripts/main.php:
```diff
@@ -6,45 +6,78 @@
+                    <label for="sort_invoice">
+                        <?php echo $this->kga['lang']['ext_invoice']['sortOrder'] ?>
+                    </label>
+                    <select name="sort_invoice">
```
id missing to connect the label with the select box

simonschaufi commented on extensions/ki_invoice/private_func.php:
```diff
@@ -200,4 +200,46 @@ function invoice_get_data($start, $end, $projects, $filter_cleared, $short_form)
+        $precision = 1.0;
+    }
+
+    return floor($value / $precision + 0.5) * $precision;
```
can i has unit tests for this please?

kevinpapst commented:
THIS IS STILL WORK IN PROGRESS.
Please do not merge until I comment again!

kevinpapst commented:
Ready for review :-)

simonschaufi commented:
unit tests still fail because of undefined constant. please check travis for more info

kevinpapst commented:
Argh, forgot to push the changed bootstrap.php - cool that we have unit tests ;-) even though only the tests are broken.
I will add that file tonight and then create a new pull request with cleaned commit history.

simonschaufi commented:
no, please no new pull request. just make another commit and don't think about the history...

kevinpapst commented:
Okay, here you go. Finalized PR

simonschaufi commented:
I did some heavy testing and found the following problems:
* when I render the 2016 template, i get an error: `
Fehler beim Rendern der Rechnung: item '<text:span' is not an existing key in the array.`
* when I render my own 2016 template, i get the following error: `Fehler beim Rendern der Rechnung: item 'fDuration' is not an existing key in the array.`

I'm not sure where the first error comes from but the second one has definitely something to do with this PR.

kevinpapst commented:
Thanks for your detailed testing and feedback! I will check everything and come back tomorrow.

kevinpapst commented:
Okay, rechecked everything...
I removed the fDuration originally and renamed it to "duration", which would break the currently used and customized user-templates out there, so I re-added the field and added "duration" field with the duration in seconds.
The broken 2016 seems to come from the change of the fDuration field. I reverted it to your version, as my OO might be the reason for the break. 
Also added "start" and "end" fields as timestamps to fix #420.

You can re-test now.

simonschaufi commented:
template variables should definitely also be documented. Please add this to the documentation repo as well

simonschaufi commented on language/en.php:
```diff
@@ -453,12 +454,18 @@
+        'failure' => 'Failed rendering invoice: %s',
+        'sortOrder' => 'Order',
+        'orderDateAsc' => 'Date - oldest first',
+        'orderDateDesc' => 'Datum - newest first',
```
this should be "Date"

kevinpapst replied:
Eagle-eye :+1:

kevinpapst commented:
I will add them to the documentation later this day and send in a PR for that. Please merge them together then.

simonschaufi commented:
@kevinpapst documentation of new variables is still missing!

kevinpapst commented:
Yes, I need to test and add them to an invoice template first, thanks for remembering!
